bjs2-75844

tanufa25 opened the pull request:
bank task is done

Commit:
bjs2-75844
bank task is done

Prigozhaeva-Anastasia commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Account.java:
```diff
@@ -0,0 +1,58 @@
+@Slf4j
+@Getter
+public class Account {
+    private final UUID id;
```
Все супер, ты молодец)

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Account.java:
```diff
@@ -0,0 +1,58 @@
+    public boolean withdraw(double amount) {
+        lock.lock();
+        try {
+            if (balance < amount) {
```
а если из 1000 рублей на счете я хочу перевести всю сумму?

20cognosce replied:
почему строгое неравенство

tanufa25 replied:
Не понимаю, что не так. Если баланс меньше, чем запрашиваемая сумма, то вывожу лог, что недостаточно средств и выхожу с false. А иначе (то есть баланс равен или больше суммы перевода) происходит перевод.

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Account.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public double getBalance() {
+        lock.lock();
+        try {
+            return balance;
+        } finally {
+            lock.unlock();
+        }
```
что это даёт?)))

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,61 @@
+        if (fromAccount.getId().equals(toAccount.getId())) {
+            log.warn("There the same account for sender and receiver");
+            return false;
+        }
```
валидации лучше отдельно выносить и по одной строчке их вызывать

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,61 @@
+
+        try {
+            if (fromAccount.withdraw(amount)) {
+                toAccount.deposit(amount);
```
зачем мы два раза локаем? внутри withdraw и deposit ты итак локи вызываешь, не понимаю

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,61 @@
+        secondLock.getLock().lock();
+
+        try {
+            if (fromAccount.withdraw(amount)) {
```
не выполнять модифицирующие операции внутри условий

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public double getTotalBalance() {
+        double value = accountMap.values().stream()
+                .mapToDouble(Account::getBalance)
+                .sum();
```
это не потокобезопасно, ни values, ни stream не синхронизированы

по-хорошему в этот момент нужно всю мапу лочить

Commit:
bjs2-75844
Locks inside Account were removed.
Lock of map during counting total balance is added.
Validations are added in new methods.

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,83 @@
+        Account toAccount = accountMap.get(to);
+
+        boolean isAnyAccountNull = isAnyAccountNull(fromAccount, toAccount);
+        if (!isAnyAccountNull) {
```
тут можно сразу в if потому что название метода понятное

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,83 @@
+            log.warn("One of the account is null, from = {}, to = {}", fromAccount, toAccount);
+            result = false;
+        }
+        return result;
```
только тут перепутаны true / false случаи
"Какой-то из аккаунтов null?" - ты возвращаешь true, когда надо false

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,83 @@
+
+    private boolean isEqualAccounts(Account fromAccount, Account toAccount) {
+        boolean result = true;
+        if (fromAccount.getId().equals(toAccount.getId())) {
```
objects.equals и тоже true/false перепутано

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,83 @@
+        secondLock.getLock().lock();
+
+        try {
+            boolean isWithdrawPossible = fromAccount.withdraw(amount);
```
не possible, а уже successful, там же самое действие выполняется

20cognosce replied:
лучше тогда переименовать в tryWithdraw, или вообще отказать от boolean и перейти на void и ошибку кидать

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,83 @@
+
+    public double getTotalBalance() {
+        double value;
+        synchronized (accountMap) {
```
Это бесполезно в текущем варианте... Ты ведь только в этом методе смотришь на этот монитор, больше нигде. Получается что ты его захватила, но всем остальным потокам всё равно, они переводы выполняют точно так же в этот момент. 

И только ты получила баланс счета в этом стриме, и прибавила его к общей сумме, он тут же обновился другим потоком

20cognosce replied:
Нужно сделать так, чтобы transfer не мог выполняться, пока мы считаем getTotalBalancе

20cognosce replied:
наверно лучше ввести атомик, и его флаг проверять в transfer, плюс сделать такую логику, чтобы перед началом подсчета мы дождались всех переводов и заняли лок

можно в цикле захватить локи всех аккаунтов, вот так наверно проще всего

20cognosce reviewed (changes requested):
Запутанная задача, я проконсультируюсь с её авторами, но давай доделаем

Commit:
bjs2-75844
Boolean methods with check are changed for void.
All accounts are locked during counting total balance

20cognosce commented on src/main/java/school/faang/multithreading/concurrency/bjs2_75844/bank/Bank.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    public double getTotalBalance() {
+        double value;
+        accountMap.values().stream()
+                .sorted(Comparator.comparing(Account::getId))
+                .forEach(account -> account.getLock().lock());
+        try {
+            value = accountMap.values().stream()
+                    .mapToDouble(Account::getBalance)
+                    .sum();
+        } finally {
+            accountMap.values().stream()
+                    .sorted(Comparator.comparing(Account::getId).reversed())
+                    .forEach(account -> account.getLock().unlock());
+        }
+        return Math.round(value * 100.0) / 100.0;
+    }
```
теперь с точки зрения блокировок всё корректно.
но на сортировки много ресурсов тратиться, может можно оптимизировать эти итерации
